src/test/app/AMM_test.cpp

```diff
@@ -4132,6 +4132,341 @@ struct AMM_test : public jtx::AMMTest
+                    if (i == 0 || i == 2)
+                    {
+                        env(offer(ed, ETH(400), USD(400)), txflags(tfPassive));
+                        env.close();
+                    }
+                    if (i > 0)
+                        amm = std::make_shared<AMM>(
+                            env, ed, USD(1'000), ETH(1'000));
+                    env(pay(carol, bob, USD(100)),
+                        path(~USD),
+                        sendmax(ETH(500)),
+                        txflags(tfPartialPayment));
+                    env.close();
+                    // CLOB and AMM, AMM is not selected
+                    if (i == 2)
+                    {
+                        BEAST_EXPECT(amm->expectBalances(
+                            USD(1'000), ETH(1'000), amm->tokens()));
+                    }
+                    BEAST_EXPECT(expectLine(env, bob, USD(2'100)));
+                    q[i] = Quality(Amounts{
+                        ETH(2'000) - env.balance(carol, ETH),
+                        env.balance(bob, USD) - USD(2'000)});
+                }
+                // CLOB is better quality than AMM
+                BEAST_EXPECT(q[0] > q[1]);
+                // AMM is not selected with CLOB
+                BEAST_EXPECT(q[0] == q[2]);
+            }
+            // Offer crossing: AMM has the same spot price quality
+            // as CLOB's offer and can't generate a better quality offer.
+            // The transfer fee in this case doesn't change the CLOB quality
+            // because the quality adjustment is ignored for the offer crossing.
+            for (auto i = 0; i < 3; ++i)
+            {
+                Env env(*this);
+                prep(env, rates.first, rates.second);
+                std::shared_ptr<AMM> amm;
+                if (i == 0 || i == 2)
+                {
+                    env(offer(ed, ETH(400), USD(400)), txflags(tfPassive));
+                    env.close();
+                }
+                if (i > 0)
+                    amm =
+                        std::make_shared<AMM>(env, ed, USD(1'000), ETH(1'000));
+                env(offer(alice, USD(400), ETH(400)));
+                env.close();
+                // AMM is not selected
+                if (i > 0)
+                {
+                    BEAST_EXPECT(amm->expectBalances(
+                        USD(1'000), ETH(1'000), amm->tokens()));
+                }
+                if (i == 0 || i == 2)
+                {
+                    // Fully crosses
+                    BEAST_EXPECT(expectOffers(env, ed, 0));
+                    BEAST_EXPECT(expectOffers(env, alice, 0));
+                }
+                // Fails to cross because AMM is not selected
+                else
+                {
+                    BEAST_EXPECT(expectOffers(
+                        env, alice, 1, {Amounts{USD(400), ETH(400)}}));
+                    BEAST_EXPECT(expectOffers(env, ed, 0));
+                }
```

Comment:
Nit: Both blocks have `BEAST_EXPECT(expectOffers(env, ed, 0));`. It makes sense to move it outside of the `if/else`.